x/perp/v2/keeper/liquidate.go

```diff
@@ -125,56 +132,65 @@ func PrettyLiquidateResponse(
 }
 
 /*
-	liquidate allows to liquidate the trader position if the margin is below the
-
+liquidate allows to liquidate the trader position if the margin is below the
 required margin maintenance ratio.
 
 args:
   - liquidator: the liquidator who is executing the liquidation
   - pair: the asset pair
   - trader: the trader who owns the position being liquidated
 
-ret:
+returns:
   - liquidatorFee: the amount of coins given to the liquidator
-  - perpEcosystemFundFee: the amount of coins given to the ecosystem fund
+  - insuranceFundFee: the amount of coins given to the ecosystem fund
   - err: error
+  - event: pointer to a typed event (proto.Message). The 'event' value
+    exists when the liquidation fails and is nil when the liquidation succeeds.
 */
 func (k Keeper) liquidate(
 	ctx sdk.Context,
 	liquidator sdk.AccAddress,
 	pair asset.Pair,
 	trader sdk.AccAddress,
-) (liquidatorFee sdk.Coin, ecosystemFundFee sdk.Coin, err error) {
+) (liquidatorFee sdk.Coin, insuranceFundFee sdk.Coin, err error) {
+	// eventLiqFailed exists when the liquidation fails and is nil when the
+	// liquidation succeeds.
+
 	market, err := k.Markets.Get(ctx, pair)
 	if err != nil {
-		_ = ctx.EventManager().EmitTypedEvent(&types.LiquidationFailedEvent{
+		eventLiqFailed := &types.LiquidationFailedEvent{
 			Pair:       pair,
 			Trader:     trader.String(),
 			Liquidator: liquidator.String(),
 			Reason:     types.LiquidationFailedEvent_NONEXISTENT_PAIR,
-		})
-		return sdk.Coin{}, sdk.Coin{}, types.ErrPairNotFound
+		}
+		_ = ctx.EventManager().EmitTypedEvent(eventLiqFailed)
+		err = fmt.Errorf("%w: pair: %s", types.ErrPairNotFound, pair)
```

Comment:
Let's use `Wrapf` instead of creating a new error, so we can preserve error codes in the response.

Reply:
addressed by 4f854da4a8c575c5c2bc7263a14e6642c6778cfe